File: src/velvet/structs/DisjointSet.java
package velvet.structs;

public class DisjointSet {

    private final int[] parents, ranks;

    public DisjointSet(int size){
        parents = new int[size];
        for(int i = 0; i < size; i++){ parents[i] = i; }
        ranks = new int[size];
    }

    private int find(int a){
        if(parents[a] != a){
            parents[a] = find(parents[a]);
        }
        return parents[a];
    }

    public boolean sameSet(int a, int b){
        return find(a) == find(b);
    }

    public boolean union(int a, int b){
        int ra = find(a);
        int rb = find(b);

        if(ra == rb){ return false; }

        if(ranks[rb] > ranks[ra]){
            int temp = ra;
            ra = rb;
            rb = temp;
        }

        parents[rb] = ra;
        if(ranks[ra] == ranks[rb]){
            ranks[ra]++;
        }

        return true;
    }
}

File: src/velvet/util/structs/DisjointSet.kt
package velvet.util.structs

class DisjointSet(size: Int) {

    private val parents: MutableList<Int> = MutableList(size){ it }
    private val ranks: MutableList<Int> = MutableList(size){ 0 }

    private fun find(a: Int): Int {
        if (parents[a] != a) {
            parents[a] = find(parents[a])
        }
        return parents[a]
    }

    fun sameSet(a: Int, b: Int): Boolean {
        return find(a) == find(b)
    }

    fun union(a: Int, b: Int): Boolean {
        var ra = find(a)
        var rb = find(b)
        if (ra == rb) {
            return false
        }
        if (ranks[rb] > ranks[ra]) {
            val temp = ra
            ra = rb
            rb = temp
        }
        parents[rb] = ra
        if (ranks[ra] == ranks[rb]) {
            ranks[ra]++
        }
        return true
    }
}